### PR TITLE
[RISCV] Add Sched classes for vector crypto instructions

### PR DESCRIPTION
**Pull request by michaelmaitland**
The vector crypto instructions may have different scheduling behavior
    compared to VALU operations. Instead of using scheduling resources that
    describe VALU operations, we give these instructions their own
    scheduling resources. This is similar to what we did for Zb* instructions.

    The sifive-p670 has vector crypto, so we model behavior for these instructions
    in the P600SchedModel. The numbers are based off of measurements collected
    internally. These numbers are a bit old and new measurements show that they may
    not be fully accurate. It is likely that we will refine these numbers in a
    follow up patch(s) based on new measurements.

    This PR is stacked on #89256.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVInstrInfoZvk.td [hunk 1]
```diff
@@ -57,8 +55,7 @@ multiclass VROR_IV_V_X_I<string opcodestr, bits<6> funct6>
   def I : RVInstIVI_VROR<funct6, (outs VR:$vd),
               (ins VR:$vs2, uimm6:$imm, VMaskOp:$vm),
               opcodestr # ".vi", "$vd, $vs2, $imm$vm">,
-         Sched<[WriteVIALUI_WorstCase, ReadVIALUV_WorstCase,
-                ReadVMask]>;
+          SchedUnaryMC<"WriteVROR", "ReadVROR">;
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrInfoZvk.td:58]: Established convention is a different Write based on the operand type. See `WriteVIALUV`, `WriteVIALUX`, and `WriteVIALUI`

File: llvm/lib/Target/RISCV/RISCVInstrInfoZvk.td [hunk 2]
```diff
@@ -24,11 +24,9 @@ def tuimm5 : RISCVOp, TImmLeaf<XLenVT, [{return isUInt<5>(Imm);}]>;
 let hasSideEffects = 0, mayLoad = 0, mayStore = 0 in {
 multiclass VCLMUL_MV_V_X<string opcodestr, bits<6> funct6> {
   def V  : VALUVV<funct6, OPMVV, opcodestr # "." # "vv">,
-           Sched<[WriteVIALUV_WorstCase, ReadVIALUV_WorstCase,
-                  ReadVIALUV_WorstCase, ReadVMask]>;
+           SchedBinaryMC<"WriteVCLMUL", "ReadVCLMUL", "ReadVCLMUL">;
   def X  : VALUVX<funct6, OPMVX, opcodestr # "." # "vx">,
-           Sched<[WriteVIALUX_WorstCase, ReadVIALUV_WorstCase,
-                  ReadVIALUX_WorstCase, ReadVMask]>;
+           SchedBinaryMC<"WriteVCLMULH", "ReadVCLMULH", "ReadVCLMULH">;
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrInfoZvk.td:29]: This is the entry for vclmul.vx and vclmulh.vx. Why using WriteVCLMULH?

File: llvm/lib/Target/RISCV/RISCVInstrInfoZvk.td [hunk 3]
```diff
@@ -24,11 +24,9 @@ def tuimm5 : RISCVOp, TImmLeaf<XLenVT, [{return isUInt<5>(Imm);}]>;
 let hasSideEffects = 0, mayLoad = 0, mayStore = 0 in {
 multiclass VCLMUL_MV_V_X<string opcodestr, bits<6> funct6> {
   def V  : VALUVV<funct6, OPMVV, opcodestr # "." # "vv">,
-           Sched<[WriteVIALUV_WorstCase, ReadVIALUV_WorstCase,
-                  ReadVIALUV_WorstCase, ReadVMask]>;
+           SchedBinaryMC<"WriteVCLMUL", "ReadVCLMUL", "ReadVCLMUL">;
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrInfoZvk.td:27]: This is for vclmul.vv and vclmulh.vv.

File: llvm/lib/Target/RISCV/RISCVScheduleZvk.td [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+
+/// Zvbc extension
+defm "" : LMULSchedWrites<"WriteVCLMUL">;
+defm "" : LMULSchedWrites<"WriteVCLMULH">;
+
+/// Zvkb extension
+defm "" : LMULSchedWrites<"WriteVANDN">;
+defm "" : LMULSchedWrites<"WriteVBREV8">;
+defm "" : LMULSchedWrites<"WriteVREV8">;
+defm "" : LMULSchedWrites<"WriteVROL">;
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVScheduleZvk.td:26]: I don't think we need separate VROL and VROR. But I do think we need RotateV, RotateX, and RotateI.

File: llvm/lib/Target/RISCV/RISCVInstrInfoZvk.td [hunk 5]
```diff
@@ -340,28 +420,104 @@ multiclass VPseudoUnaryV_V<LMULInfo m> {
...
+    defm "" : VPseudoUnaryV_V<m>,
+              SchedUnary<"WriteVCPOPV", "ReadVCPOPV", mx, forceMergeOpRead=true>;
   }
 }
 
 multiclass VPseudoVWALU_VV_VX_VI<Operand ImmType> : VPseudoVWALU_VV_VX {
   foreach m = MxListW in {
     defm "" : VPseudoBinaryW_VI<ImmType, m>,
-              SchedUnary<"WriteVIWALUV", "ReadVIWALUV", m.MX,
+              SchedUnary<"WriteVWSLLV", "ReadVWSLLV", m.MX,
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrInfoZvk.td:458]: This should be WriteVWSLLI. The non-VI instructions are getting a scheduler class from the inheriting of `VPseudoVWALU_VV_VX`. That seems odd.

File: llvm/lib/Target/RISCV/RISCVInstrInfoZvk.td [hunk 6]
```diff
@@ -340,28 +420,104 @@ multiclass VPseudoUnaryV_V<LMULInfo m> {
...
                          forceMergeOpRead=true>;
   }
 }
 
+multiclass VPseudoVANDN {
+ foreach m = MxList in {
+    defm "" : VPseudoBinaryV_VV<m>,
+            SchedBinary<"WriteVANDNV", "ReadVANDNV", "ReadVANDNV", m.MX,
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrInfoZvk.td:466]: I think we can use the existing VIALUV

**Review by topperc — APPROVED** [review]
LGTM

File: llvm/lib/Target/RISCV/RISCVScheduleZvk.td [hunk 7]
```diff
@@ -0,0 +1,208 @@
+//===- RISCVScheduleB.td - RISC-V Scheduling Definitions B -*- tablegen -*-===//
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVScheduleZvk.td:1]: Copy paste mistake here.

Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/RISCVScheduleZvk.td]: updated

File: llvm/lib/Target/RISCV/RISCVScheduleZvk.td [hunk 8]
```diff
@@ -0,0 +1,208 @@
+//===- RISCVScheduleB.td - RISC-V Scheduling Definitions Zvk -*- tablegen -===//
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVScheduleZvk.td:1]: The file name is still incorrect.

File: llvm/lib/Target/RISCV/RISCVInstrInfoZvk.td [hunk 9]
```diff
@@ -340,28 +420,111 @@ multiclass VPseudoUnaryV_V<LMULInfo m> {
...
-              SchedUnary<"WriteVIWALUV", "ReadVIWALUV", m.MX,
+              SchedUnary<"WriteVWSLLI", "ReadVWSLLV", mx,
                          forceMergeOpRead=true>;
   }
 }
 
+multiclass VPseudoVANDN {
+ foreach m = MxList in {
+    defm "" : VPseudoBinaryV_VV<m>,
+            SchedBinary<"WriteVIALUV", "ReadVIALUV", "ReadVIALUV", m.MX,
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVInstrInfoZvk.td:473]: Indent here.

File: llvm/lib/Target/RISCV/RISCVInstrInfoZvk.td [hunk 10]
```diff
@@ -340,28 +420,111 @@ multiclass VPseudoUnaryV_V<LMULInfo m> {
...
+                        forceMergeOpRead=true>;
+    defm "" : VPseudoBinaryV_VX<m>,
+            SchedBinary<"WriteVIALUX", "ReadVIALUV", "ReadVIALUX", m.MX,
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVInstrInfoZvk.td:476]: Ditto.

File: llvm/lib/Target/RISCV/RISCVInstrInfoZvk.td [hunk 11]
```diff
@@ -340,28 +420,111 @@ multiclass VPseudoUnaryV_V<LMULInfo m> {
...
+    defvar mx = m.MX;
+    defm "" : VPseudoUnaryV_V<m>,
+              SchedUnary<"WriteVREV8V", "ReadVREV8V", mx, forceMergeOpRead=true>;
+  }
+}
+
+multiclass VPseudoVROL {
+ foreach m = MxList in {
+    defm "" : VPseudoBinaryV_VV<m>,
+            SchedBinary<"WriteVRotV", "ReadVRotV", "ReadVRotV", m.MX,
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVInstrInfoZvk.td:500]: ditto.

File: llvm/lib/Target/RISCV/RISCVInstrInfoZvk.td [hunk 12]
```diff
@@ -340,28 +420,111 @@ multiclass VPseudoUnaryV_V<LMULInfo m> {
...
+                        forceMergeOpRead=true>;
+    defm "" : VPseudoBinaryV_VX<m>,
+            SchedBinary<"WriteVRotX", "ReadVRotV", "ReadVRotX", m.MX,
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVInstrInfoZvk.td:503]: ditto.

File: llvm/lib/Target/RISCV/RISCVInstrInfoZvk.td [hunk 13]
```diff
@@ -340,28 +420,111 @@ multiclass VPseudoUnaryV_V<LMULInfo m> {
...
+                        forceMergeOpRead=true>;
+  }
+}
+
+multiclass VPseudoVROR<Operand ImmType> {
+  defvar Constraint = "";
+  foreach m = MxList in {
+    defvar mx = m.MX;
+    defm "" : VPseudoBinaryV_VV<m>,
+            SchedBinary<"WriteVRotV", "ReadVRotV", "ReadVRotV", mx,
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVInstrInfoZvk.td:513]: ditto.

File: llvm/lib/Target/RISCV/RISCVScheduleZvk.td [hunk 14]
```diff
@@ -0,0 +1,208 @@
...
+
+/// Zvksed extension
+defm "" : LMULSchedWrites<"WriteVSM4KV">;
+defm "" : LMULSchedWrites<"WriteVSM4RV">;
+
+/// Zvksh extension
+defm "" : LMULSchedWrites<"WriteVSM3CV">;
+defm "" : LMULSchedWrites<"WriteVSM3MEV">;
+
+/// Define scheduler resources associated with use operands.
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVScheduleZvk.td:55]: Remove this line?

Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/RISCVScheduleZvk.td:55]: Took like from `RISCVScheduleZb.td`. At top of file we have line for def. This is line for use. Do you want me to remove both?

Comment (wangpc-pp) [reply llvm/lib/Target/RISCV/RISCVScheduleZvk.td:55]: Oh sorry, please just keep it.

**Review by wangpc-pp — APPROVED** [review]
LGTM.

**Pull request merged**